boot/bootstate20.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	// get the kernel for this bootloader
+	bootSnap, err = ebl.Kernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot locate kernel snap with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	tryKernel, tryKernelExists, err := ebl.TryKernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot locate try kernel snap with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	if tryKernelExists {
+		tryBootSnap = tryKernel
+	}
+
+	m, err := bl.GetBootVars("kernel_status")
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	trying = (m["kernel_status"] == "trying")
+
+	return bootSnap, tryBootSnap, trying, nil
+}
+
+func (s20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	var u20 *bootStateUpdate20
+	var err error
+	u20, err = newBootStateUpdate20(update, snap.TypeKernel)
+	if err != nil {
+		return nil, err
+	}
+
+	// kernel_status goes from "" -> "try" -> "trying" -> ""
+	// so if we are not in "trying" mode, nothing to do here
+	if u20.env["kernel_status"] != "trying" {
+		return u20, nil
+	}
+
+	// if kernel_status is trying, then we need to figure out what kernel we did
+	// boot from
+	sn, exists, err := u20.ebl.TryKernel()
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		// somehow we booted with kernel_status=trying, but no try-kernel
+		// symlink :-/
+		return nil, fmt.Errorf("cannot mark boot successful: kernel_status is \"trying\", but no try-kernel enabled")
+	}
+
+	// set the snap we tried to the snap from the try-kernel symlink
+	u20.snapTried = sn
+
+	// update the boot var
+	u20.toCommit["kernel_status"] = ""
+
+	return u20, nil
+}
+
+func (s20 *bootState20Kernel) setNext(nextKernel snap.PlaceInfo) (rebootRequired bool, u bootStateUpdate, err error) {
+	u20, err := newBootStateUpdate20(nil, snap.TypeKernel)
+	if err != nil {
+		return false, nil, err
+	}
+
+	currentKernel, err := u20.ebl.Kernel()
+	if err != nil {
+		return false, nil, err
+	}
+
+	kernelStatus := "try"
+	rebootRequired = true
+	// check if the next kernel snap is the same as the current kernel snap,
+	// in which case we either don't need to do anything or we just need to
+	// clear the status and not reboot
+	if nextKernel.SnapName() == currentKernel.SnapName() && nextKernel.SnapRevision() == currentKernel.SnapRevision() {
+		// If we were in anything but default ("") mode before
+		// and switched to the good core/kernel again, make
+		// sure to clean the kernel_status here. This also
+		// mitigates https://forum.snapcraft.io/t/5253
+		if u20.env["kernel_status"] == "" {
+			// already clean
+			return false, nil, nil
+		}
+
+		// clean
+		kernelStatus = ""
+		rebootRequired = false
+	} else {
+		// need to enable the symlink
+		u20.snapToTry = nextKernel
+	}
+
+	u20.toCommit["kernel_status"] = kernelStatus
+
+	return rebootRequired, u20, nil
+}
+
+type bootStateUpdate20 struct {
```

Comment:
a general remark: for 16/18 having a single bootStateUpdate16 was quite clean, here maybe having a base struct with helpers and then different ones with different commit impl for setNext vs markSuccessful might be clearer, not sure but wanted to let use consider that option, also reminder only markSuccessful need *Update chaining/folding support

Reply:
@anonymouse64 are you going to play a bit with that idea? or should I review what is there as is for now?

Reply:
Sorry I thought I had responded to this comment, but apparently I hadn't as I don't see that comment here :-(

Anyways, I would prefer to leave the structure as-is for this PR and refactor in a follow-up when I work on base snap updates.

Reply:
But to be clear, I think your suggestion is a good idea, but I'd rather land what I have now and iterate than re-factor again in this PR if that's okay